extensions/amp-story-shopping/0.1/amp-story-shopping-config.js

```diff
@@ -8,23 +16,99 @@ export class AmpStoryShoppingConfig extends AMP.BaseElement {
     super(element);
 
     /** @private {string} */
-    this.myText_ = TAG;
+    this.element_ = element;
+
+    /** @private @const {!./amp-story-store-service.AmpStoryStoreService} */
+    this.storeService_ = null;
+  }
 
-    /** @private {?Element} */
-    this.container_ = null;
+  /**
+   * Keys product data to product-ids and adds them to the store service.
+   * @private
+   * @param {!JsonObject} storyConfig
+   */
+  dispatchConfig_(storyConfig) {
+    const productIDtoProduct = {};
+
+    for (const item of storyConfig['items']) {
+      productIDtoProduct[item['product-tag-id']] = item;
+    }
+
+    this.storeService_.dispatch(Action.ADD_SHOPPING_STATE, productIDtoProduct);
+
+    //TODO: add call to validate config here. See validation I2I #36412 for implementation details
   }
 
   /** @override */
   buildCallback() {
     super.buildCallback();
-    this.container_ = this.element.ownerDocument.createElement('div');
-    this.container_.textContent = this.myText_;
-    this.element.appendChild(this.container_);
-    applyFillContent(this.container_, /* replacedContent */ true);
+    return Services.storyStoreServiceForOrNull(this.win).then(
+      (storeService) => {
+        devAssert(storeService, 'Could not retrieve AmpStoryStoreService');
+
+        this.storeService_ = storeService;
+
+        this.getConfig().then((storyConfig) => {
+          //if remote config is not valid, try using the inline config
+          if (storyConfig == null && this.element_.hasAttribute('src')) {
+            this.getInlineConfig_(this.element_.firstElementChild).then(
+              (storyInlineconfig) => {
+                this.dispatchConfig_(storyInlineconfig);
+              }
+            );
+          } else {
+            this.dispatchConfig_(storyConfig);
+          }
+        });
+      }
+    );
   }
 
   /** @override */
   isLayoutSupported(layout) {
     return layout === Layout.NODISPLAY;
   }
+
+  /**
+   * @return {!JsonObject}
+   */
+  getConfig() {
+    const configData = this.element_.hasAttribute('src')
+      ? this.getRemoteConfig_()
+      : this.getInlineConfig_(this.element_.firstElementChild);
+
+    return configData;
+  }
+
+  /**
+   * @param {!Element} child
+   * @return {!JsonObject}
+   */
+  getInlineConfig_(child) {
+    userAssert(
+      child && isJsonScriptTag(child),
+      `The ${TAG} should ` +
+        'be inside a <script> tag with type="application/json"'
+    );
+
+    const inlineJSONConfig = parseJson(child.textContent);
+
+    return Promise.resolve(inlineJSONConfig);
```

Comment:
I think this could just return the JSON. Does it need to return a Promise.resolve?

If we did that, the the getConfig chain could be simplified even more to:
```
this.getConfig()
  .then((storyConfig) => this.dispatchConfig_(storyConfig))
  .catch(() => this.dispatchConfig_(this.getInlineConfig_()));
```

Reply:
I tried that and it will not work - we are calling this with getConfig, and using.then, which requires a promise object. It would require more refactoring than the changes to getConfig you listed above.

Reply:
Oh I see now! Thank you :) 
Maybe there's an opportunity here. We only need to branch once: If remote fails, fallback to inline.
What if we remove getConfig and just do:
```
this.getRemoteConfig_()
  .then((storyConfig) => this.dispatchConfig_(storyConfig))
  // If remote fails, fallback to inline.
  .catch(() => this.dispatchConfig_(this.getInlineConfig_()));
```
and in getRemoteConfig we reject the promise if src does not exist:
```
getRemoteConfig_() {
  const src = this.element.getAttribute('src');
  if (!src) {
    return Promise.reject();
  }
 ...
```

Reply:
Yeah, this makes a lot more sense and works now! Thanks!